File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
       
        int leftMax = 0 , rightMax = 0 , left = 0 , right = height.size()-1 , res = 0;
        
        while(left <= right){
            
            if(height[left] <= height[right]){
                
                if(height[left] < leftMax) res += leftMax - height[left];
                
                else leftMax = height[left];
                
                left++;
            }
            
            else{
                
                if(height[right] < rightMax) res += rightMax - height[right];
                
                else rightMax = height[right];
                
                right--;
            }
        }
        return res;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length ;
        int rMax = 0 , lMax = 0 , l = 0 , r = n - 1 , ans = 0 ;
        while(l<=r){
            if(height[l]<=height[r]){
                if(height[l]<lMax) ans += lMax-height[l] ;
                lMax = Math.max(lMax,height[l]) ;
                l++ ;
            }else{
                if(height[r]<rMax) ans += rMax-height[r] ;
                rMax = Math.max(rMax,height[r]) ;
                r--;
            }
        }
        return ans ;
    }
}